modyn/tests/trainer_server/internal/trainer/remote_downsamplers/test_rs2_downsampling.py

```diff
@@ -0,0 +1,172 @@
+    assert downsampler._all_sample_ids == 2 * sample_ids
+    # Now it should not change anymore
+    downsampler.select_points()
+    downsampler.inform_samples(sample_ids, forward_output, target)
+    assert set(downsampler._all_sample_ids) == set(sample_ids)
+    assert len(downsampler._all_sample_ids) == 2 * len(sample_ids)
+
+
+def test_multiple_epochs_with_replacement():
+    pipeline_id = 0
+    trigger_id = 0
+    batch_size = 32
+    params_from_selector = {"replacement": True, "downsampling_ratio": 50}
+    per_sample_loss = None
+    device = "cpu"
+
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+    with torch.inference_mode(mode=(not downsampler.requires_grad)):
+        sample_ids = list(range(10))
+        data = torch.randn(10, 10)
+        target = torch.randint(0, 10, (10,))
+
+        for _ in range(3):
+            downsampler.inform_samples(sample_ids, data, target)
+            selected_ids, weights = downsampler.select_points()
+            sampled_data, sampled_target = get_tensors_subset(selected_ids, data, target, sample_ids)
+
+            assert len(set(selected_ids)) == 5
+            assert weights.shape == (5,)
+            assert all(idx in sample_ids for idx in selected_ids)
+            assert sampled_data.shape == (5, 10)
+            assert sampled_target.shape == (5,)
+
+
+def test_multiple_epochs_without_replacement():
+    pipeline_id = 0
+    trigger_id = 0
+    batch_size = 32
+    params_from_selector = {"replacement": False, "downsampling_ratio": 50}
+    per_sample_loss = None
+    device = "cpu"
+
+    downsampler = RemoteRS2Downsampling(
+        pipeline_id, trigger_id, batch_size, params_from_selector, per_sample_loss, device
+    )
+    with torch.inference_mode(mode=(not downsampler.requires_grad)):
+
+        sample_ids = list(range(10))
+        data = torch.randn(10, 10)
+        target = torch.randint(0, 10, (10,))
+
+        # Epoch 1
+        downsampler.inform_samples(sample_ids, data, target)
+        epoch1_ids, weights = downsampler.select_points()
+        sampled_data, sampled_target = get_tensors_subset(epoch1_ids, data, target, sample_ids)
+
+        assert len(set(epoch1_ids)) == 5
+        assert weights.shape == (5,)
+        assert all(idx in sample_ids for idx in epoch1_ids)
+        assert sampled_data.shape == (5, 10)
+        assert sampled_target.shape == (5,)
+
+        # Epoch 2
+        downsampler.inform_samples(sample_ids, data, target)
+        epoch2_ids, weights = downsampler.select_points()
+        sampled_data, sampled_target = get_tensors_subset(epoch2_ids, data, target, sample_ids)
+
+        assert len(set(epoch2_ids)) == 5
+        assert weights.shape == (5,)
+        assert all(idx in sample_ids for idx in epoch2_ids)
+        assert not any(idx in epoch1_ids for idx in epoch2_ids)  # No overlap across epochs
```

Comment:
How does this line absolutely pass? For each epoch we reshuffle the sample ids. Even if now we choose the second subset, the ids in the second subset can still come from the first subset in the previous epoch right? (because there is not a definite relation between the first subset in the first epoch and the second subset in the second epoch, as we shuffle the entire sample ids in between)

Reply:
Sorry, I don't understand your question. In RS2, we generate a new subset each epoch (= round in the paper). Now, with replacement, we just always get some x% subset of the IDs. When we do without replacement, we have to iterate through the entire dataset (minus some leftover depending on the size), and then restart. Imagine we have these IDs

1 2 3 4 5

now we shuffle

2 1 4 5 3

now we partition it into subsets

[2 1] [4 5] 3

first epoch will return 2 1, second epoch will return 4 5

then we will have to reshuffle

3 2 1 5 4

partition again

[3 2] [1 5] 4

third epoch will return 3 2, fourth will return 1 5

Does this answer your question?

Reply:
Ah I see.I think this line confuses me https://github.com/eth-easl/modyn/blob/240dbf49de69b27821ea39c36867aae0e3bb66bb/modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py#L65

Actually, in the `NoReplacement` case, we only need to reshuffle at https://github.com/eth-easl/modyn/blob/240dbf49de69b27821ea39c36867aae0e3bb66bb/modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py#L60 right? Can you move the shuffle statement to each `step` method?

Reply:
Sure. I am not opinionated on where we shuffle (I thought it's not an expensive operation so I can just unify w/wo replacement) but I moved it into the step functions